From 7297747a2e77e555133ebf668c587d9706c3eae7 Mon Sep 17 00:00:00 2001
From: Christian Bruel <christian.bruel@foss.st.com>
Date: Mon, 23 Sep 2024 17:12:30 +0200
Subject: [PATCH 755/841] fix(psci): Allow return to aarch32 smc caller

Entering PSCI from SMC(ARM32) call, the ESR mode must be checked
as the upper EL layer might still be in aarch64. This is the case
when U-boot aarch64/EL2 starts a aarch32/EL1 kernel.
Check return mode from ESR.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 78d176623ffcb2f9d744a790a86ae4981fb752a9 fix(psci): Allow return to aarch32 smc caller]
Signed-off-by: Christian Bruel <christian.bruel@foss.st.com>
Change-Id: I83bec98476b8c39c375015032b190aebd2354f40
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/406275
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 lib/psci/psci_common.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/lib/psci/psci_common.c b/lib/psci/psci_common.c
index bc973dce3..2d07af1db 100644
--- a/lib/psci/psci_common.c
+++ b/lib/psci/psci_common.c
@@ -828,6 +828,7 @@ static int psci_get_ns_ep_info(entry_point_info_t *ep,
 	u_register_t ep_attr, sctlr;
 	unsigned int daif, ee, mode;
 	u_register_t ns_scr_el3 = read_scr_el3();
+	u_register_t ns_esr_el3 = read_esr_el3();
 	u_register_t ns_sctlr_el1 = read_sctlr_el1();
 
 	sctlr = ((ns_scr_el3 & SCR_HCE_BIT) != 0U) ?
@@ -847,9 +848,15 @@ static int psci_get_ns_ep_info(entry_point_info_t *ep,
 
 	/*
 	 * Figure out whether the cpu enters the non-secure address space
-	 * in aarch32 or aarch64
+	 * in aarch32 or aarch64 from ESR first, or SCR elsewhere.
 	 */
-	if ((ns_scr_el3 & SCR_RW_BIT) != 0U) {
+	if (EC_BITS(ns_esr_el3) == EC_AARCH32_SMC) {
+		mode = MODE32_svc;
+
+		daif = DAIF_ABT_BIT | DAIF_IRQ_BIT | DAIF_FIQ_BIT;
+
+		ep->spsr = SPSR_MODE32((uint64_t)mode, entrypoint & 0x1, ee, daif);
+	} else if ((ns_scr_el3 & SCR_RW_BIT) != 0U) {
 
 		/*
 		 * Check whether a Thumb entry point has been provided for an
-- 
2.47.2

